Blog post: Analysing Quarkus Native startup RSS

galderz opened the pull request:
A new blog post explaining how we analysed Quarkus Native startup RSS consumption.

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+
+=== `ObjectAllocationInNewTLAB` JFR event
+
+Another improvement since we carried out this investigation can be found in the area of tooling.
+At the time, the only tooling we could really use here was `perf` or `eBPF`, which is only available in Linux.
+Additions to the emitted JFR events for native applications in latest GraalVM Community versions include the `jdk.ObjectAllocationInNewTLAB` event,
+which is the JFR the event that would have highlighted discrepancies in the Quarkus versions compared above.
+Unfortunately, the Quarkus versions tested here are not compatible with this GraalVM version,
+but we can give it a go with the latest Quarkus version to see what comes out.
```
```suggestion
As newer versions of GraalVM implement https://github.com/oracle/graal/issues/5410[more and more JFR events] in native mode it also becomes easier to analyse Quarkus application performance. 
For instance the recent release of GraalVM for JDK 17/20 add support for the `jdk.ObjectAllocationInNewTLAB` event which would have highlighted the discrepancies in the Quarkus versions compared above.
Unfortunately, the Quarkus versions tested here are not compatible with this GraalVM version,
but we can give it a go with the latest Quarkus version to see what comes out.
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+Unfortunately, the Quarkus versions tested here are not compatible with this GraalVM version,
+but we can give it a go with the latest Quarkus version to see what comes out.
+
+First, build the Quarkus native with JFR:
```
```suggestion
First, we build the native Quarkus application with JFR support:
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+
+To obtain `jdk.ObjectAllocationInNewTLAB` events,
+a custom JFC configuration file is required.
+`jfr configure` can be used to generate it:
```
```suggestion
We use `jfr configure` to generate it:
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+$ $JAVA_HOME/bin/jfr configure jdk.ObjectAllocationInNewTLAB#enabled=true --output newtlab.jfc
+----
+
+Then, start Quarkus native executable with the necessary JFR flags:
```
```suggestion
Then, we start the Quarkus native executable with the necessary JFR flags:
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+$ $JAVA_HOME/bin/jfr print --events ObjectAllocationInNewTLAB --stack-depth 100 recording.jfr
+----
+
+Amongst the output you will find events such as the following,
+which are related to the flamegraph stacktraces that we highlighted above:
```
```suggestion
Inspecting the output we see events related to the flamegraph stacktraces that we highlighted above:
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+
+[source,bash]
+----
+jdk.ObjectAllocationInNewTLAB {
```
In this example we don't see the call to `OSCommittedMemoryProvider.allocate()` though, could you perhaps use an event including this as well, or is this something not shown in the JFR event for some reason?

galderz replied:
The discrepancy is due where the JFR event is emitted vs when the `mmap` system call happens. The JFR event gets emitted after the `slowPathNew*` method has finished, and this method is the one that calls `OSCommittedMemoryProvider.allocate()` before the event fires.

I think the JFR event is fine as is, and don't think there's a need to add a new event. However, what I can maybe do is focus the user on searching for `slowPathNew` methods, and that way the similarities between the two approaches should be smaller. WDYT?

zakkak replied:
Yes, either explain the difference, or move the focus to `slowPathNew` (explaining why it's important), it might be obvious to people familiar with GC, but I guess it's not for most people :)

galderz replied:
The new updates contain a detailed explanation of the slow/fast allocation paths.

zakkak reviewed (changes requested):
Great piece of information Galder, thanks.

I added a number of suggestions and comments to hopefully further improve it.

roberttoyonaga commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,228 @@
+If you focus the flamegraph on that method,
+and then click on `Search` at the top right and type `allocate`,
+you can observe which of the `sys_enter_mmap` system calls are related to native heap memory allocation.
+These allocations are for heap chunks, each of which is 1MB,
```
Maybe its worth mentioning it's no longer 1MB and has since been changed to 512KB

galderz replied:
Added a small comment around the upcoming change to reduce it.

roberttoyonaga commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,228 @@
+image::rss-after-mmap-allocate.png[heap chunk allocation related mmap calls for Quarkus 2.13.4.Final]
+
+The increase in object allocation at runtime startup between these two Quarkus versions,
+combined with the idiosyncrasies of how thread local allocation works in SubstrateVM,
```
Maybe it's worth elaborating on "idiosyncrasies". For example, explaining that Hotspot has dynamic TLAB resizing whereas SVM has a one-size-tries-to-fit-all approach, which is why the issue only pops up in native mode even though both modes experience more allocations.

galderz replied:
This has been solved in the latest update. In expanding https://github.com/quarkusio/quarkusio.github.io/pull/1755#discussion_r1272268418, I've explained how the slow/fast path work and the TLA logic in SubstrateVM. I've then expanded this paragraph further.

roberttoyonaga commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,228 @@
+----
+
+After shutting down Quarkus,
+we can use https://adoptium.net/jmc[Adoptium's Eclipse Mission Control],
```
Just in case user's are unfamiliar with JMC, it might be good to mention that you can see user-friendly flamegraphs of this JFR event (similar to what you obtained with perf and eBPF).

galderz replied:
I'm adding a JMC screenshot but I don't think the flamegraphs in JMC add anything productive in this case. It assumes each object allocation is a different flamegraph, because a different object is allocated, and show doesn't show them together in the same way you saw with perf/eBPF.

roberttoyonaga replied:
Do you mean you are only able to see the stacktrace of a single event at a time? If you click on the event itself or highlight a series of events, you can view them all together.  Is this what you mean? 
![image](https://github.com/quarkusio/quarkusio.github.io/assets/35396843/9178478b-a3b8-472c-b53e-8401e1346d60)

The behavior you describe (viewing one class at a time) is also possible though:
![image](https://github.com/quarkusio/quarkusio.github.io/assets/35396843/52d13fe5-baa4-4b3b-8d84-1dc1ce488f06)

galderz replied:
Ah ok. I couldn't get to see that initially. I had to click back on the "Event browser" and then select the event type to see them all together. I'll try to add that to the blog.

roberttoyonaga replied:
Yup, I agree the UI is a little clunky. I do wish there was a way to quickly "select all". But as far as I know there isn't :(

galderz replied:
If you select multiple events at the same time, e.g. all events coming from `main` thread, the flamegraph expands to cover all of those, which as expected all root at the `ApplicationImpl.doStart()` method. I added this screenshot to the blog post.

galderz commented:
@roberttoyonaga @zakkak Pushed a commit to address all remaining feedback:

* Added detailed explanation around SubstrateVM thread local allocation.
* Explained differences between fast and slow allocation paths.
* Moved focus on flamegraph from `allocate` method to those methods that are prefixed with `slowPathNew`. Adjusted flamegraph screenshots accordingly.
* Provide further explanation on the summary of differences between HotSpot and SubstrateVM.
* Added JMC screenshot showing a new tlab jfr event relevant to this blog post.

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,257 @@
+
+We describe the object allocation as fast when the thread has a heap chunk cached and there's enough space for the object.
+We describe the object allocation as slow when the thread needs to request a new heap chunk via the `mmap` system call to satisfy the object allocation.
+This latter type of allocations are the most interesting ones for us,
```
```suggestion
Slow allocations are the most interesting in this case,
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,257 @@
+We describe the object allocation as fast when the thread has a heap chunk cached and there's enough space for the object.
+We describe the object allocation as slow when the thread needs to request a new heap chunk via the `mmap` system call to satisfy the object allocation.
+This latter type of allocations are the most interesting ones for us,
+because they give us an rough indication of which allocations are pushing the number of heap chunks up,
```
```suggestion
because they give us a rough indication of which allocations are pushing the number of heap chunks up,
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,257 @@
+This latter type of allocations are the most interesting ones for us,
+because they give us an rough indication of which allocations are pushing the number of heap chunks up,
+and therefore are pushing the RSS consumption up.
+In SubstrateVM, slow allocation path stacktraces contain `com.oracle.svm.core.genscavenge.ThreadLocalAllocation` class invocations to methods with `slowPathNew` prefix.
```
```suggestion
In SubstrateVM, stacktraces of the slow allocation path contain invocations to methods defined in the `com.oracle.svm.core.genscavenge.ThreadLocalAllocation` class that contain the `slowPathNew` prefix in their name .
```

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,257 @@
+
+The areas of interest in the flamegraphs are those stacktraces that contain `ApplicationImpl.doStart()` method call,
+which are invoked from the `main` thread.
+We want to find out how many slow path allocations that are visible in these stacktraces.
```
```suggestion
We want to find out how many slow path allocations are visible in these stacktraces.
```

zakkak reviewed (changes requested):
Thanks for the changes Galder.

There are some typos for which I added suggestions, please have a look.

galderz commented:
@roberttoyonaga @zakkak I've pushed some new changes that introduce a similar flamegraph view with JMC focused on the new TLAB allocations. This data, obtained with Quarkus 3.2.1.Final (plus latest GraalVM community) raises some potential questions. I'm on the move and cannot evaluate this further but I can do that in ~2 weeks.

zakkak reviewed (approved):
@galderz the new changes look good.

Note that there are still some unresolved comments though in the PR.
Just making sure they didn't slip through.

galderz commented:
@zakkak Hmm, I thought I had handled all comments. What's still missing?

zakkak commented:
> @zakkak Hmm, I thought I had handled all comments. What's still missing?

Annoyingly, github hides unresolved comments because there are a lot of comments in total...

![image](https://github.com/quarkusio/quarkusio.github.io/assets/1435395/6c92af45-006a-4390-b7a5-2e9148e760c4)

You can see all the comments by either clicking "Load more..." or scrolling through [the "Files changed" tab](https://github.com/quarkusio/quarkusio.github.io/pull/1755/files).

galderz commented:
Ah I missed those! Thanks :)

galderz commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+combined with the idiosyncrasies of how thread local allocation works in SubstrateVM,
+is what caused this increase in RSS consumption.
+
+Since then, improvements were made to the components that caused the extra object allocation.
```
Right, I left that information out.

The idea is to look at the components that appear in the stacktrace leading to the slow allocation and see what changes have happened in those. This is not a perfect science since it could happen that a slow allocation results from somewhere that is not responsible for the increase (by luck), but it gives some good hints. In this case it was changes in how the SmallRye configuration was handled at runtime.

Shall I add something along those lines?

galderz commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+In this blog post you will learn how
+https://quarkus.io/guides/native-reference#rss[these Quarkus native RSS consumption debugging techniques]
+were used to diagnose this issue.
+You will also learn about recent updates in GraalVM that make doing this kind of analysis more convenient.
```
Yeah that should be updated. @roberttoyonaga would you like to have a go at updating it? The idea would be to go through the [native reference guide](https://quarkus.io/guides/native-reference) and see where and how JFR can provide alternative methods to debug some of the issues highlighted here.

galderz commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+
+Following
+https://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html[Brendan Gregg’s Memory Leak (and Growth) Flame Graphs guide] on a Linux environment,
+we were able to narrow down the issue to the shape of the flamegraphs for the `mmap` system calls.
```
Initially I wasn't sure where the RSS increase was coming from. When I found the article, I went one by one trying all the system calls that were presented there and learning how to generate flamegraphs for each. When I did that I realised that the `mmap` flamegraph showed a lot of calls which then I started digging into the SVM logic that was triggering it. This flamegraph was also the one that presented differences between the two Quarkus versions.

I could have probably used `strace` to see which system calls are the ones that are called the most, and then use the article to see how to generate flamegraphs for those that are used the most.

Shall I elaborate this in the blog post?

zakkak replied:
Yes, please. No need to say much, just say that RSS can only be increased through system calls and that you started looking for them and that's how you ended up focusing on `mmap`.

galderz commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,230 @@
+https://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html[Brendan Gregg’s Memory Leak (and Growth) Flame Graphs guide] on a Linux environment,
+we were able to narrow down the issue to the shape of the flamegraphs for the `mmap` system calls.
+To obtain the flamegraphs, we capture the `sys_enter_mmap` system call events with `perf record`,
+and then we generate flamegraphs with the recorded data:
```
Let in the blog post I'm explaining the role of `mmap` in SVM heap allocation when I talk about slow/fast allocations. Shall I move that explanation upfront?

zakkak replied:
No, I think that the current content with the possible addition of https://github.com/quarkusio/quarkusio.github.io/pull/1755#discussion_r1274588860 is enough. Thanks!

galderz commented:
@zakkak I've pushed a commit with further changes. It should address all remaining feedback.

zakkak commented on _posts/2023-07-24-native-startup-rss-troubleshooting.adoc:
```diff
@@ -0,0 +1,223 @@
+but due to the non-deterministic nature of object allocation,
+the stacktrace just happens to be the one that caused a new heap chunk to be allocated.
+However, if you see multiple stacktraces in the same area,
+that's probably a good hint that something in that stacktrace needs to be investigated.
```
```suggestion
that's probably a good hint that something in that area needs to be investigated.
```

zakkak reviewed (approved):
LGTM, except for a typo (I think).

Thank you for addressing all comments.

galderz commented:
No probs @zakkak. I'll go ahead and squash commits now.
